Mathlib/LinearAlgebra/QuadraticForm/Basic.lean

```diff
@@ -587,341 +601,359 @@ def linMulLin (f g : M →ₗ[R] R) : QuadraticForm R M where
+theorem toQuadraticMap_eq_zero {B : BilinMap R M N} :
+    B.toQuadraticMap = 0 ↔ B.IsAlt :=
+  QuadraticMap.ext_iff
+#align bilin_form.to_quadratic_form_eq_zero LinearMap.BilinMap.toQuadraticMap_eq_zero
 
 end Semiring
 
 section Ring
 
-open QuadraticForm
-
-variable [CommRing R] [AddCommGroup M] [Module R M]
-variable {B : BilinForm R M}
+variable [CommRing R] [AddCommGroup M] [AddCommGroup N] [Module R M] [Module R N]
+variable {B : BilinMap R M N}
 
 @[simp]
-theorem toQuadraticForm_neg (B : BilinForm R M) : (-B).toQuadraticForm = -B.toQuadraticForm :=
+theorem toQuadraticMap_neg (B : BilinMap R M N) : (-B).toQuadraticMap = -B.toQuadraticMap :=
   rfl
-#align bilin_form.to_quadratic_form_neg LinearMap.BilinForm.toQuadraticForm_neg
+#align bilin_form.to_quadratic_form_neg LinearMap.BilinMap.toQuadraticMap_neg
 
 @[simp]
-theorem toQuadraticForm_sub (B₁ B₂ : BilinForm R M) :
-    (B₁ - B₂).toQuadraticForm = B₁.toQuadraticForm - B₂.toQuadraticForm :=
+theorem toQuadraticMap_sub (B₁ B₂ : BilinMap R M N) :
+    (B₁ - B₂).toQuadraticMap = B₁.toQuadraticMap - B₂.toQuadraticMap :=
   rfl
-#align bilin_form.to_quadratic_form_sub LinearMap.BilinForm.toQuadraticForm_sub
-
-theorem polar_toQuadraticForm (x y : M) : polar (toQuadraticForm B) x y = B x y + B y x := by
-  simp only [toQuadraticForm_apply, add_assoc, add_sub_cancel_left, add_apply, polar, add_left_inj,
-    add_neg_cancel_left, map_add, sub_eq_add_neg _ (B y y), add_comm (B y x) _]
-#align bilin_form.polar_to_quadratic_form LinearMap.BilinForm.polar_toQuadraticForm
+#align bilin_form.to_quadratic_form_sub LinearMap.BilinMap.toQuadraticMap_sub
+
+theorem polar_toQuadraticMap (x y : M) : polar (toQuadraticMap B) x y = B x y + B y x := by
+  simp only [polar, toQuadraticMap_apply, map_add, add_apply, add_assoc, add_comm (B y x) _,
+    add_sub_cancel_left, sub_eq_add_neg _ (B y y), add_neg_cancel_left]
+#align bilin_form.polar_to_quadratic_form LinearMap.BilinMap.polar_toQuadraticMap
+
+theorem polarBilin_toQuadraticMap : polarBilin (toQuadraticMap B) = B + flip B :=
+  LinearMap.ext₂ polar_toQuadraticMap
+
+@[simp] theorem _root_.QuadraticMap.toQuadraticMap_polarBilin (Q : QuadraticMap R M N) :
+    toQuadraticMap (polarBilin Q) = 2 • Q :=
+  QuadraticMap.ext fun x => (polar_self _ x).trans <| by simp
+
+theorem  _root_.QuadraticMap.polarBilin_injective (h : Invertible (2 : R)) :
+    Function.Injective (polarBilin : QuadraticMap R M N → _) :=
+  fun Q₁ Q₂ h₁₂ => QuadraticMap.ext fun x => by
+  have e1 : 2 • Q₁ x = 2 • Q₂ x := by
+    simpa using DFunLike.congr_fun (congr_arg toQuadraticMap h₁₂) x
+  calc
+    Q₁ x = (⅟(2 : R) * 2) • Q₁ x := by rw [invOf_mul_self', one_smul]
+    _ = ⅟(2 : R) • ((2 : R) • Q₁ x) := by rw [mul_smul]
+    _ = ⅟(2 : R) • (2 • Q₁ x) := by rw [two_smul, ← two_smul ℕ]
+    _ = ⅟(2 : R) • (2 • Q₂ x) := by rw [e1]
+    _ = ⅟(2 : R) • ((2 : R) • Q₂ x) := by rw [two_smul ℕ, two_smul]
+    _ = Q₂ x := by rw [← mul_smul, invOf_mul_self', one_smul]
```

Comment:
Here is my attempt at making this simpler:

```suggestion
theorem _root_.QuadraticMap.polarBilin_injective (h : IsUnit (2 : R)) :
    Function.Injective (polarBilin : QuadraticMap R M N → _) := by
  intro Q₁ Q₂ h₁₂
  apply h.smul_left_cancel.mp
  rw [show (2 : R) = (2 : ℕ) by rfl]
  simp_rw [← nsmul_eq_smul_cast R, ← QuadraticMap.toQuadraticMap_polarBilin]
  exact congrArg toQuadraticMap h₁₂
```

I prefer `IsUnit` to `Invertible` here. `Invertible` would be acceptable if it was a typeclass argument.